src/util/thread.h

```diff
@@ -235,5 +235,190 @@ class UpdateThread : public Thread
+{
+public:
+	ThreadPool(const std::string &name, size_t thread_count = 1) :
+		m_name(name),
+		m_threads(thread_count < 0 ? 0 : thread_count)
+	{
+		if (thread_count < 0)
+			thread_count = 0;
+
+		for (size_t i = 0; i < thread_count; ++i) {
+			spawnThread();
+		}
+	}
+	~ThreadPool()
+	{
+		stopAllThreadsWait();
+	}
+
+	void stopAllThreads()
+	{
+		for (size_t i = 0; i < m_threads.size(); ++i) {
+			if (m_threads[i]->isRunning()) {
+				m_threads[i]->stop();
+				stopThread();
+			}
+		}
+	}
+
+	void stopAllThreadsWait()
+	{
+		stopAllThreads();
+		size_t i = runningThreadCount();
+		while (i--) {
+			m_threads[i]->wait();
+		}
+	}
+
+	void setRunningThreadCount(size_t count)
+	{
+		if (count < 0)
+			count = 0;
+
+		const size_t thread_count = runningThreadCount();
+		if (thread_count == count)
+			return;
+
+		if (thread_count < count) {
+			for (size_t i = thread_count; i < count; ++i)
+				spawnThread();
+		}
+		else /* (m_thread_count > count) */ {
+			for (size_t i = count; i < thread_count; ++i)
+				stopThread();
+		}
+	}
+
+	size_t threadCount() const
+	{
+		return m_threads.size();
+	}
+
+	size_t runningThreadCount() const
+	{
+		size_t running = 0;
+		for (size_t i = 0; i < m_threads.size(); ++i) {
+			if (m_threads[i]->isRunning())
+				++running;
+		}
+		return running;
+	}
+
+	inline bool isRunning() const
+	{
+		return runningThreadCount() > 0;
+	}
+
+	ThreadClass &getThread(size_t id)
+	{
+		return *m_threads[id];
+	}
+
+	void enqueue(const MessageType &msg)
+	{
+		m_msgqueue.push_back(ThreadPoolMessage<MessageType>(false, msg));
+		m_sem.post();
+	}
+
+protected:
+	virtual ThreadClass *makeThread() = 0;
+
+	int findFreeIndex(bool invert = false)
+	{
+		int free_idx = -1;
+		for (size_t i = 0; i < m_threads.size(); ++i) {
+			if ((!m_threads[i]->isRunning()) ^ invert) {
+				free_idx = i;
+				break;
+			}
+		}
+		return free_idx;
+	}
+
+	void spawnThread()
+	{
+		int free_idx = findFreeIndex();
+		ThreadClass *thr = makeThread();
+		if (free_idx == -1) {
+			m_threads.push_back(thr);
+		}
+		else {
+			delete m_threads[free_idx];
+			m_threads[free_idx] = thr;
+		}
+	}
+
+	bool stopThread()
+	{
+		if (m_threads.empty())
+			return false;
+
+		if (findFreeIndex(true) != -1) {
+			m_msgqueue.push_back(ThreadPoolMessage<MessageType>(true, MessageType()));
+			m_sem.post();
+			return true;
+		}
+		return false;
+	}
+
+	std::string m_name;
+	std::vector<ThreadClass*> m_threads;
+	Semaphore m_sem;
+	MutexedQueue< ThreadPoolMessage<MessageType> > m_msgqueue;
+};
+
+template<typename MessageType>
+class PoolableUpdateThread : public Thread
+{
+public:
+	PoolableUpdateThread(const std::string &name, Semaphore *sem,
+		MutexedQueue< ThreadPoolMessage<MessageType> > *msgqueue) :
+		Thread(name + "Update"),
+		m_sem(sem),
+		m_msgqueue(msgqueue) {}
+	~PoolableUpdateThread() {}
+
+	void *run()
+	{
+		DSTACK(FUNCTION_NAME);
+		BEGIN_DEBUG_EXCEPTION_HANDLER
+
+		while (!stopRequested()) {
+			m_sem->wait();
+
+			if (stopRequested()) break;
```

Comment:
@kwolekr does not like this at all. Better put the break on a separate line, and I think its also in the code style